packages/core/src/storages/request_queue_v2.ts

```diff
@@ -100,17 +95,52 @@ export class RequestQueue extends RequestProvider {
             id: queueOperationInfo.requestId,
             isHandled: queueOperationInfo.wasAlreadyHandled,
             uniqueKey: queueOperationInfo.uniqueKey,
+            forefront: queueOperationInfo.forefront,
             hydrated: null,
             lockExpiresAt: null,
         });
     }
 
+    /**
+     * @inheritDoc
+     */
+    override async addRequest(
+        requestLike: Source,
+        options: RequestQueueOperationOptions = {},
+    ): Promise<RequestQueueOperationInfo> {
+        const result = await super.addRequest(requestLike, options);
+        if (!result.wasAlreadyPresent && options.forefront) {
+            this.shouldCheckForForefrontRequests = true;
+        }
+        return result;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    override async addRequests(
+        requestsLike: Source[],
+        options: RequestQueueOperationOptions = {},
+    ): Promise<BatchAddRequestsResult> {
+        const result = await super.addRequests(requestsLike, options);
+        for (const request of result.processedRequests) {
+            if (!request.wasAlreadyPresent && options.forefront) {
+                this.shouldCheckForForefrontRequests = true;
+            }
```

Comment:
You can break once you found one, no reason to keep searching
```suggestion
            if (!request.wasAlreadyPresent && options.forefront) {
                this.shouldCheckForForefrontRequests = true;
                break;
            }
```